pkg/migrate/cstor/blockdevice_correction.go

```diff
@@ -0,0 +1,217 @@
+			return err
+		}
+	}
+
+	for i, bdname := range spcObj.Spec.BlockDevices.BlockDeviceList {
+		if correctedBD[bdname] != "" {
+			spcObj.Spec.BlockDevices.BlockDeviceList[i] = correctedBD[bdname]
+		}
+	}
+	delete(spcObj.Annotations, string(apis.OpenEBSDisableReconcileKey))
+	_, err = spc.NewKubeClient().Update(spcObj)
+	return err
+}
+
+func (c *CSPCMigrator) correctCSPBDs(cspObj apis.CStorPool) error {
+	podList, err := c.KubeClientset.CoreV1().Pods(c.OpenebsNamespace).
+		List(metav1.ListOptions{
+			LabelSelector: "openebs.io/cstor-pool=" + cspObj.Name,
+		})
+	if err != nil {
+		return err
+	}
+	if len(podList.Items) != 1 {
+		return errors.Errorf("failed to get csp pod expected 1 got %d", len(podList.Items))
+	}
+	devLinks, err := c.getDevlinks(podList.Items[0].Name)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get devlinks for csp %s", cspObj.Name)
+	}
+	devLinkBDMap := map[string]string{}
+	index := 0
+	for _, group := range cspObj.Spec.Group {
+		for _, bd := range group.Item {
+			devLinkBDMap[devLinks[index]] = bd.Name
+			correctedBD[bd.Name] = ""
+			index = index + 1
+		}
+	}
+	correctCSPBD := map[string]string{}
+	hostName := podList.Items[0].Spec.NodeSelector["kubernetes.io/hostname"]
+	for devlink, bdname := range devLinkBDMap {
+		bdObj, err := c.OpenebsClientset.OpenebsV1alpha1().
+			BlockDevices(c.OpenebsNamespace).
+			Get(bdname, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		if bdObj.Status.State == "Active" && bdObj.Status.ClaimState == "Claimed" {
+			nodes, err := c.KubeClientset.CoreV1().Nodes().
+				List(metav1.ListOptions{
+					LabelSelector: "kubernetes.io/hostname=" + hostName,
+				})
+			if err != nil {
+				return err
+			}
+			if len(nodes.Items) == 1 {
+				continue
+			}
+		}
+		newBD, err := c.findBDforDevlink(devlink, hostName)
+		if err != nil {
+			return err
+		}
+		correctCSPBD[bdname] = newBD
+		correctedBD[bdname] = newBD
+	}
+
+	newCSPObj := cspObj.DeepCopy()
+
+	for i, rg := range newCSPObj.Spec.Group {
+		for j, bd := range rg.Item {
+			if correctCSPBD[bd.Name] != "" {
+				newCSPObj.Spec.Group[i].Item[j].Name = correctCSPBD[bd.Name]
+				newCSPObj.Spec.Group[i].Item[j].DeviceID = ""
+			}
+		}
+	}
+	data, err := GetPatchData(cspObj, newCSPObj)
+	if err != nil {
+		return err
+	}
+	_, err = csp.KubeClient().
+		Patch(cspObj.Name, types.MergePatchType, data)
+
+	return err
+}
+
+func (c *CSPCMigrator) findBDforDevlink(devlink, hostname string) (string, error) {
+	bds, err := c.OpenebsClientset.OpenebsV1alpha1().
+		BlockDevices(c.OpenebsNamespace).
+		List(metav1.ListOptions{
+			LabelSelector: "kubernetes.io/hostname=" + hostname,
+		})
+	if err != nil {
+		return "", err
+	}
+	for _, bd := range bds.Items {
+		for _, devLinks := range bd.Spec.DevLinks {
+			for _, link := range devLinks.Links {
+				if strings.Contains(devlink, link) && bd.Status.State == "Active" &&
+					bd.Status.ClaimState == "Unclaimed" {
+					return bd.Name, nil
+				}
+			}
+		}
```

Comment:
There will be cases where block device exists even without devlinks but they will have `bd.Spec.Path` need to handle such cases. 
cc: @akhilerm

Reply:
the same logic can be applied to BD with no devlinks right?

Reply:
Yes